homeassistant/components/optoma_projector/switch.py

```diff
@@ -0,0 +1,108 @@
+"""
+Use Optoma's Web interface to control the projector.
+
+For more details about this component, please refer to the documentation
+at https://home-assistant.io/components/switch.optoma_projector/
+"""
+import logging
+from urllib import parse
+import hashlib
+import requests
+from bs4 import BeautifulSoup
+import voluptuous as vol
+
+from homeassistant.components.switch import (SwitchDevice, PLATFORM_SCHEMA)
+from homeassistant.const import (
+    CONF_HOST, CONF_NAME, CONF_USERNAME, CONF_PASSWORD)
+import homeassistant.helpers.config_validation as cv
+
+_LOGGER = logging.getLogger(__name__)
+
+DEFAULT_NAME = 'Optoma Projector'
+ICON = 'mdi:projector'
+DEFAULT_USERNAME = 'admin'
+DEFAULT_PASSWORD = 'admin'
+PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend({
+    vol.Required(CONF_HOST): cv.string,
+    vol.Optional(CONF_NAME, default=DEFAULT_NAME): cv.string,
+    vol.Optional(CONF_USERNAME, default=DEFAULT_USERNAME): cv.string,
+    vol.Optional(CONF_PASSWORD, default=DEFAULT_PASSWORD): cv.string,
+})
+
+
+def setup_platform(hass, config, add_entities, discovery_info=None):
+    """Return an Optoma Projector."""
+    host = config.get(CONF_HOST)
+    name = config.get(CONF_NAME)
+    username = config.get(CONF_USERNAME)
+    password = config.get(CONF_PASSWORD)
+
+    add_entities([OptomaSwitch(host, name, username, password)], True)
+
+
+class OptomaSwitch(SwitchDevice):
+    """Represents an Optoma Projector as a switch."""
+
+    def __init__(self, host, name, username, password, **kwargs):
+        """Init of the Optoma projector."""
+        self.url = 'http://'+host
+        self._name = name
+        self.username = username
+        self.password = password
+        self.session = requests.session()
+        self._state = False
+        self._login()
+
+    def _login(self):
+        response = self.session.get(parse.urljoin(self.url, '/login.htm'))
```

Comment:
Please extract the device specific interface code, I/O and data parsing, to a standalone library published on pypi.

https://developers.home-assistant.io/docs/en/creating_platform_code_review.html#4-communication-with-devices-services